flang/lib/Lower/Bridge.cpp

```diff
@@ -336,10 +336,42 @@ class FirConverter : public Fortran::lower::AbstractConverter {
         llvm::None);
   }
 
-  void createVar(const Fortran::semantics::Symbol &sym) override final {
-    Fortran::lower::AggregateStoreMap fakeMap;
-    Fortran::lower::pft::Variable var(sym, /*global=*/false);
-    instantiateVar(var, fakeMap);
+  bool createVar(const Fortran::semantics::Symbol &sym) override final {
+    auto getShape = [](Fortran::lower::SymbolBox sb) {
+      return sb.match(
+          [&](const Fortran::lower::SymbolBox::FullDim &arr)
+              -> llvm::ArrayRef<mlir::Value> { return arr.getExtents(); },
+          [&](const Fortran::lower::SymbolBox::CharFullDim &arr)
+              -> llvm::ArrayRef<mlir::Value> { return arr.getExtents(); },
+          [&](const Fortran::lower::SymbolBox::Box &arr)
+              -> llvm::ArrayRef<mlir::Value> {
+            return arr.getExplicitExtents();
```

Comment:
The `SymbolBox::Box` (or `fir::BoxValue` in the ExtentedValue classification) case is a bit more complex.

`getExplicitExtents` won't get the extents from `SymbolBox::Box` in the general case. A variable represented by a `SymbolBox::Box` in lowering is a variable that is a descriptor (and is not an allocatable/pointer). `getExplicitExtents` will return extents if this variable has an explicit shape. Variable that are explicit shape and are described by a descriptor are quite rare (Right now, I think this will be limited to explicit shape polymorphic entities, which we do not lower yet).

The local host associated temp created from a `SymbolBox::Box` will in most case not be a descriptor. In most cases, the root variable is tracked in lowering as a descriptor because it is not contiguous, which will not be the case of the local copy (I see one case where the local copy will still probably need a descriptor, which is polymorphism, but we are not there yet).

So right now, it means two things for you here:

 - You need to use `Fortran::lower::getExtents()` from `FIRBuilder.h` to read the extents from the descriptor and use that in the alloca. You can replace `getShape` by it if you directly work on `hexv`, it will work on all categories. Note that this will generate code reading the original descriptor, so you may have to think when this need to be generated from an OpenMP point of view (before or inside the region).

- In the `fir::BoxValue` case (same as `SymbolBox::Box` for extended values), `fir::substbase` cannot be used, because you are not replacing a descriptor by a descriptor. The temp will be a `fir::CharArrayBoxValue` or `fir::ArrayBoxValue`.

All that to say, I would probably rather organize the code as:
```c++
auto allocate = [&](llvm::ArrayRef<mlir::Value> shape, llvm::ArrayRef<mlir::Value> typeParams) -> mlir::Value {
    return builder->allocateLocal(
        genLocation(sym.name()), genType(sym), mangleName(sym),
        toStringRef(sym.GetUltimate().name()), shape, typeParams),
        sym.GetUltimate().attrs().test(Fortran::semantics::Attr::TARGET));
};

fir::ExtendedValue hexv = hsb.toExtendedValue();
auto exval = hexv.match(
  [&](const fir::BoxValue &box) -> fir::ExtendedValue {
    const auto* type = sym.GetType();
    if (type && type->IsPolymorphic())
      TODO(loc, "create polymorphic host associated copy");
    // Create a contiguous temp with the same shape and length as
    // the original variable described by a fir.box.
    auto extents = Fortran::lower::getExtents(*builder, loc, hexv);
    if (box.isDerivedWithLengthParameters())
      TODO(loc, "get length parameters from derived type BoxValue");
    if (box.isCharacter()) {
     auto len = Fortran::lower::readCharLen(*builder, loc, box);
     auto temp = allocate(extents, {len}); 
     return fir::CharArrayBoxValue{temp, len, extents};
    }
    return fir::ArrayBoxValue{allocate(extents, {}), extents};
  },
  [&](const fir::MutableBoxValue &box) -> fir::ExtendedValue {
    // Allocate storage for a pointer/allocatble descriptor.
    // No shape/lengths to be passed to the alloca.
    return fir::MutableBoxValue(allocate({}, {}), box.nonDeferredLenParams(), {});
  },
  [&](const auto &) -> fir::ExtendedValue {
     auto temp = allocate(Fortran::lower::getExtents(*builder, loc, hexv),
                                           fir::getTypeParams(hexv));
     return fir::substBase(hexv, v);
  });
 return bindSymbol(sym, exval);
```

This code might not compile, I did not try, it's more of an illustration.
Note that you might have to update `allocateLocal` to deal with the elimination of uneeded extents/type parameters. fir.alloca does not want length and extents argument that are already constant in the type being allocated.

Reply:
Thanks @jeanPerier for the detailed comments and code. The code worked with minor syntax changes.

I had to add some code for the allocatable case conversion to extended value.

The cases that I tried did not show any issues with allocateLocal. Could you give a hint on the cases I should try?

> Note that this will generate code reading the original descriptor, so you may have to think when this need to be generated from an OpenMP point of view (before or inside the region)
Currently allocas and related code are hoisted to the entry block of the closest outlineable function (like parallel, task).

Reply:
Something where extents/lengths of private variables are compile time constants:

```fortran
subroutine array(x, c)
integer :: i
integer :: x(100)
character(5) :: c
!$omp parallel private(x, c)
x = i
c = "abcde"
!$omp end parallel
print *, x
end subroutine
```

What will most likely happen is that character case will create a FIR verification complaining about passing a length parameter for a fir.alloca of character type whose length is already in the type.

For the array case, I think you will not get verification error, but the alloca will allocate a the square of the size in the end (e.g. `fir.alloca !fir.array<100xi32>, %c100` which allocates storage for a hundred `!fir.array<100xi32>`, which is not the intention here).

Reply:
> For the array case, I think you will not get verification error, but the alloca will allocate a the square of the size in the end (e.g. `fir.alloca !fir.array<100xi32>, %c100` which allocates storage for a hundred `!fir.array<100xi32>`, which is not the intention here).

@jeanPerier I handled these cases in the allocate lambda (as follows) by checking whether the types have constant shape or constant length. 
Are there more cases to consider?
Did you suggest putting these in the allocateLocal function?

```
    auto allocate = [&](llvm::ArrayRef<mlir::Value> shape,
                        llvm::ArrayRef<mlir::Value> typeParams) -> mlir::Value {
      mlir::Type symType = genType(sym);
      bool hasConstantShape = false;
      bool hasConstantLength = false;
      if (auto arrTy = symType.dyn_cast<fir::SequenceType>())
        hasConstantShape = arrTy.hasConstantShape();
      if (auto charTy = symType.dyn_cast<fir::CharacterType>())
        hasConstantLength = charTy.hasConstantLen();
      return builder->allocateLocal(
          loc, symType, mangleName(sym), toStringRef(sym.GetUltimate().name()),
          hasConstantShape ? llvm::SmallVector<mlir::Value>{} : shape,
          hasConstantLength ? llvm::SmallVector<mlir::Value>{} : typeParams,
          sym.GetUltimate().attrs().test(Fortran::semantics::Attr::TARGET));
    };
```

Reply:
> Are there more cases to consider?
Not that I can think of.

> Did you suggest putting these in the allocateLocal function?

Yes, I think it would be great to have this done directly in allocateLocal so that we do not have to deal with this here and there. There are already some helpers in `lib/Lower/FIRBuilder.cpp` to deal with this (`elideExtentsAlreadyInType`/ `elideLengthsAlreadyInType`).

Reply:
Thanks @jeanPerier. Moved the code to allocateLocal and using  elideExtentsAlreadyInType/elideLengthsAlreadyInType.